### ISSUE TITLE
prov/tcp RFC: CQ spin polling without poll() overhead

### ISSUE THREAD

**Issue by sydidelot** [ofiwg/libfabric#6745]
Hi,

I am currently running some performance measurements and profiling with the TCP provider (latest master).

The TCP provider relies on epoll()/poll() to monitor the socket events and progress pending socket operations.
It also needs to update the POLLOUT event in the poll set if outgoing buffers are waiting for the given socket. 
This scenario also happens when the user application wants to spin for completion (CQ configured with `FI_WAIT_NONE`).

I understand that poll is the best solution to minimize the CPU utilization while waiting for socket events.
However, some users (I am part of theem :)) may want to minimize the impact of poll() in the fast path. 

I did run some latency benchmarks with fabtest (command line is `./fi_msg_pingpong -S all -c spin`) where I compare libfabric *with* and *without* calling poll(). Here is the corresponding patch if you want to give it a try:
```
diff --git a/prov/tcp/src/tcpx_cq.c b/prov/tcp/src/tcpx_cq.c
index 0828b6194..94ea25555 100644
--- a/prov/tcp/src/tcpx_cq.c
+++ b/prov/tcp/src/tcpx_cq.c
@@ -37,6 +37,7 @@
 
 #define TCPX_DEF_CQ_SIZE (1024)
 
+#define POLL_ENABLED
 
 void tcpx_cq_progress(struct util_cq *cq)
 {
@@ -63,14 +64,22 @@ void tcpx_cq_progress(struct util_cq *cq)
                 * of length 0, there's no additional data to read, so failing
                 * to progress can result in application hangs.
                 */
+#ifndef POLL_ENABLED
+               tcpx_progress_tx(ep);
+               tcpx_progress_rx(ep);
+#else
                if (ofi_bsock_readable(&ep->bsock) ||
                    (ep->cur_rx.handler && !ep->cur_rx.entry))
                        tcpx_progress_rx(ep);
+#endif
 
+#ifdef POLL_ENABLED
                (void) tcpx_update_epoll(ep);
+#endif
                fastlock_release(&ep->lock);
        }
 
+#ifdef POLL_ENABLED
        nfds = (wait_fd->util_wait.wait_obj == FI_WAIT_FD) ?
               ofi_epoll_wait(wait_fd->epoll_fd, wait_contexts,
                              MAX_POLL_EVENTS, 0) :
@@ -96,6 +105,7 @@ void tcpx_cq_progress(struct util_cq *cq)
                fastlock_release(&ep->lock);
        }
 unlock:
+#endif
        cq->cq_fastlock_release(&cq->ep_list_lock);
 }
```

The results are available in the following Google spreadsheet:
https://docs.google.com/spreadsheets/d/1cA3OEsqkt6bHK9hUiahQEWr39V2ppBE5NiosT_mMvgI/edit?usp=sharing

In a nutshell, the version without poll() gives a speedup of ~1.18 for small messages on my setup (localhost interface is used).
I think it would make sense to expose to the user application the ability to completely disable poll(), especially on small setups where only a few endpoints need to be progressed.
Also, I think it could be useful to support switching on/off poll() after the CQ was created: the application may want to start with poll() disabled when the CQ has a few endpoints bound to it and switch on poll() support as soon as a certain threshold is reached. 

What do you think?
Thank you!

**Comment by shefty** [comment]
Another thought is that we could also look at tracking hot sockets and calling those directly.  I question the benefit, since we'd still need to poll on the other sockets -- maybe a small latency benefit handling received data.

**Comment by shefty** [comment]
I need to understand the spreadsheet better.  The y-axis shows usecs/transfer, so lower should be better, correct?  The lower is listed as poll.  Given the test, this is for a single connection.  So I would expect that spinning would outperform poll -- not sure why the results suggest the opposite.  

I think these data points would be useful:

- time to call read on an empty socket (POLLIN false)
- times to call poll on 1 to N sockets

That should help determine the read to poll threshold.

**Comment by aakefbs** [comment]
I guess this is asking to add support for io_uring.

https://github.com/frevib/io_uring-echo-server/blob/io-uring-feat-fast-poll/benchmarks/benchmarks.md

**Comment by sydidelot** [comment]
Thanks a lot for your inputs, @shefty. That's really helpful! 

Sorry for my stupidity: I have inverted the labels. The version without poll() is effectively the one that performs better (lower value for usec/xfer is better). 

It is clear from your comments that the threshold largely depends on the performance of send(), recv() and epoll(), which may be influenced by the OS or the CPU architecture.

@aakefbs raises an important point where io_uring will be a game changer for polling as it will be limiting the number of syscalls.
In fact, I have been working on adding a support of io_uring to the tcp provider and this work led me to benchmark the polling function. I would like to release that patch upstream as soon as I get the approval from DDN.

**Comment by shefty** [comment]
io_uring is in serious need of usable documentation.  Based on reading through the man pages, examples, and general documentation, I agree with @aakefbs that we could use it to replace epoll/poll, and I would anticipate it giving better performance.  We maybe even want a new progress loop specific to io_uring.

We could also use io_uring to perform asynchronous socket send/recv calls.  The performance benefit of that isn't as obvious to me.  I think we can queue sends and possibly improve small message bw -- assuming user->kernel transitions can introduce pipeline stalls.  I can't tell if ZEROCOPY is supported by io_uring.  I don't yet know how async socket recv calls would work through io_uring, to guess if there's any potential benefit.

**Comment by sydidelot** [comment]
@shefty here are some resources you may find useful to understand io_uring:
- Lord of the io_uring: https://unixism.net/loti/
- liburing source code: https://github.com/axboe/liburing and more especially the following header file that contains all the function in the API: https://github.com/axboe/liburing/blob/master/src/include/liburing.h

> Will also need to figure out where the distros and CSPs are wrt supporting these.

I am using Ubuntu 20.10 (5.8.0-50-generic) and liburing package from apt: io_uring works reliably with this setup.

> We could also use io_uring to perform asynchronous socket send/recv calls. The performance benefit of that isn't as obvious to me. 

Performing a socket operation requires the following steps:
- Allocate a Submission Queue Entry (SQE) from the ring (io_uring_get_sqe()) and fill it with your corresponding socket operation (e.g., io_uring_prep_sendmsg()). Optionally you can set a context with io_uring_sqe_set_data(), which you can later retrieve when the operation completes.
- The SQE is submitted to the kernel when io_uring_submit() is executed. 
- Check Completion Queue Entry (CQE) completions (i.e., SQEs that completed) with io_uring_peek_cqe() (or any other equivalent function). After CQE processing is done, call io_uring_cqe_seen() to return the CQE to the ring.
- That's all :)

I see several benefits of using io_uring:
- Checking for completion is a lightweight operation as it happens in user-space (no syscall). You don't need to poll the socket to check if the data is ready anymore.
- io_uring supports batched operations: you can allocate multiple SQEs and submit all of them with a single call to io_uring_submit().


> I don't yet know how async socket recv calls would work through io_uring, to guess if there's any potential benefit.

io_uring implements a feature called IORING_FEAT_FAST_POLL. With that feature you don't have to poll the socket to check if incoming data is ready. You just post a receive buffer asynchronously and a CQE is generated when there is data ready in that buffer. More information is available here: https://github.com/frevib/io_uring-echo-server/issues/5

> I can't tell if ZEROCOPY is supported by io_uring

I'm not familiar with ZEROCOPY for socket operations but I found a discussion about ZEROCOPY in the io-uring mailing list: https://www.spinics.net/lists/io-uring/msg05258.html

I hope my message will help you to understand io_uring better.

**Comment by shefty** [comment]
I read through all of those links yesterday, along with reading example code.  There's a lot of conceptual discussion, but API details are left as an exercise to the reader.  Socket ZEROCOPY support is documented about as well.

I look at io_uring as *possibly* helping in 3 higher-level areas.  The first is as a replacement to poll/epoll.  But I don't see that io_uring can be used as a drop-in replacement for epoll.  That is, I can't use it to tell if a socket is readable or writeable.  I can only post send/recv requests and wait for those to complete.

Which brings me to receive handling...  The general tcp flow is to recv the tcp header, read the message length, recv additional headers, find the target buffer, and recv data into the target buffer.  I can't tell what the behavior is if I post an 8k receive to an io_uring.  Does that operation not complete until 8k has been received, or will the buffer be partially filled?  We need the latter behavior.

For send handling, I have the same question.  If I post a 1MB send, will the operation complete only after all 1 MB has been transferred, or will it complete after the outbound socket buffers have been filled.  We want the first behavior here, which is opposite of the receive case.

I need to know how io_uring impacts the behavior of the socket calls in order to design around it.

Example: If a recv posted for X bytes does not complete until X bytes are available to be read, then the flow would be: post 8 byte recv, wait for cqe, post recv to get rest of headers, wait for cqe, post recv to get user's data, wait for cqe.  For small messages, that will be significantly slower than what we have now.  So, we would only want to use io_uring to receive larger user data.  But going this route means the code still relies on epoll to see if a socket is readable.  And since the main advantage of io_uring is to shave a couple microseconds of latency, if we're only using it for larger transfers, I doubt we'd see any benefits.

If a recv posted for X bytes can complete as soon as any data is available, then io_uring is more usable, at least on the receive side.  But now we run into design issues on the send side.

**Comment by shefty** [comment]
If a send is posted to a non-writable socket, will the send defer until at least some amount of data is transferred?  If so, then based on what you mentioned, I think we can make use of it without significant redesign.

I agree that an io_uring per ep is the most straightforward option.  Maybe placing the io_uring with the domain might be more efficient, if we can avoid iterating over a cq's ep list

**Comment by sydidelot** [comment]
> If a send is posted to a non-writable socket, will the send defer until at least some amount of data is transferred?
Yes, this is what I observed in my prototype: the send doesn't immediately return 0 and its completion is defered until some amout of data is transferred.

It would be nice if we could avoid iterating over a cq's ep list, but we will need to ensure that the EP always has async operations waiting for completion. If the EP progress function returns without starting at least one async operation, we will need to manually progress the EP or register the EP's socket FD to a poll set for a retry.

**Comment by sydidelot** [comment]
@shefty, regarding your previous comment:
> Maybe placing the io_uring with the domain might be more efficient, if we can avoid iterating over a cq's ep list

I have one concern if we place the `io_uring` with the domain: multiple threads polling different CQ from the same domain would be serialized. The application I'm developing creates multiple polling threads per network device and each thread has its own CQ and set of endpoints. One io_uring per domain would cause a high contention between the polling threads.

Placing the `io_uring` in the CQ would be optimal but it cannot work if the user application binds the EP to different CQs for TX and RX operations. For example, completing TX operations would sometimes require to poll the RX CQ.

I'm a bit puzzled here.

**Comment by shefty** [comment]
When an ep is bound to different cq's for tx and rx, the current code ends up needing the socket associated with both the tx and rx epoll fd's.  Both epoll fd's end up needing to be updated, for example, when a socket is no longer writeable and we need to check for pollout to continue progress.  So, we don't have per cq thread isolation.

We can get per domain thread isolation.  But the app would need to know to create endpoints in separate domains.  (FWIW, rxm doesn't do this, but does use a single cq.  I expect a single cq per ep would be the common case.)

As a thought, if we created a 'progress_domain' (names are hard) as an internal object, we could associate the io_uring and epoll fd there.  If the endpoints use a single cq, the progress_domain would be with the cq.  If the endpoint has 2 cq's, it's progress_domain would be associated with those cq's (or just placed at the domain level to make things simpler).

To me, this sounds reasonable in theory, but likely impossible to implement.  :)  To implement we would need the ability to merge or migrate between progress_domains on the fly while they are actively in use.  In practice, I doubt we'd realistically hit that flow however.

**Comment by sydidelot** [comment]
Closing the issue as PR #8645 allows for progressing TCP communications without poll()/epoll() overhead.
